### PR TITLE
Allowlisting for delegation pools

### PR DESCRIPTION
**Pull request by alexfilip2**
### Description

This implements AIP-31: https://github.com/aptos-foundation/AIPs/blob/main/aips/aip-31.md
The owner of a delegation pool would be able to enforce that only particular accounts can join the pool as delegators. The owner can also evict existing ineligible delegators by unlocking their entire stake.

### Test Plan
Module UTs validating core functionality and invariants.

### CODE REVIEW

**Comment by junkil-park** [comment]
Can you generate `.md` files and include them in this PR?

File: aptos-move/framework/move-stdlib/sources/configs/features.move [hunk 1]
```diff
@@ -392,6 +392,16 @@ module std::features {
         is_enabled(CONCURRENT_FUNGIBLE_ASSETS) && aggregator_v2_api_enabled()
     }
 
+    /// Whether delegators allowlisting for delegation pools is supported.
+    /// Lifetime: transient
+    const DELEGATION_POOL_ALLOWLISTING: u64 = 50;
```
Comment (junkil-park) [inline aptos-move/framework/move-stdlib/sources/configs/features.move:397]: Can you add this feature flag to `aptos_features.rs` and `feature_flags.rs`?

File: aptos-move/framework/aptos-framework/sources/delegation_pool.move [hunk 2]
```diff
@@ -325,6 +339,13 @@ module aptos_framework::delegation_pool {
         effective_after_secs: u64,
     }
 
+    /// Tracks a delegation pool's allowlist of delegators.
+    /// If allowlisting is enabled, existing delegators are not implicitly allowlisted and they can be individually
+    /// evicted later by the pool owner.
+    struct DelegationPoolAllowlisting has key {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/delegation_pool.move:345]: Is it possible to move this functionality into a separate module? In 2023, there have been enough improvements to Move and the Aptos Framework to allow for much more modular designs that how this and the stake module were written in 2022 + early 2023.

File: aptos-move/framework/aptos-framework/sources/delegation_pool.move [hunk 3]
```diff
@@ -1813,6 +2034,9 @@ module aptos_framework::delegation_pool {
     #[test_only]
     const COMMISSION_CHANGE_DELEGATION_POOL: u64 = 42;
 
+    #[test_only]
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/delegation_pool.move:2037]: Also start separating the tests into a separate file under a tests/ directory

File: aptos-move/framework/aptos-framework/sources/delegation_pool.move (hunk 2 above)
Comment (alexfilip2) [reply aptos-move/framework/aptos-framework/sources/delegation_pool.move:345]: I agree, the allowlist resource and events could be moved to a separate module.
`assert_delegator_allowlisted` can be entirely implemented there and be externally available for `add_stake` and `reactivate_stake`.
However, operations that require the pool's ownership capability (especially `evict_delegator`) would have to be partially implemented in `delegation_pool.move` and forward the request to `delegation_pool_allowlist.move` to update the allowlist.
Validating the ownership entirely in `delegation_pool_allowlist.move` would require importing `delegation_pool.move` which becomes a circular dependency. We could save the owner address in `DelegationPoolAllowlisting` when enabling allowlisting and then update it if `DelegationPoolOwnership` should ever be transferable.

What would be the recent improvements of the Aptos Framework that could help us with this separation?

File: aptos-move/framework/aptos-framework/sources/delegation_pool.move (hunk 3 above)
Comment (junkil-park) [reply aptos-move/framework/aptos-framework/sources/delegation_pool.move:2037]: We will separate tests in bulk on another PR.

**Comment by junkil-park** [comment]
This PR migrates to https://github.com/aptos-labs/aptos-core/pull/12213.

**Pull request closed without merging**